include/ginkgo/core/matrix/batch_dense.hpp

```diff
@@ -0,0 +1,372 @@
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/mtx_io.hpp>
+#include <ginkgo/core/base/range_accessors.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+namespace batch {
+namespace matrix {
+
+
+/**
+ * Dense is a batch matrix format which explicitly stores all values of the
+ * matrix in each of the batches.
+ *
+ * The values in each of the batches are stored in row-major format (values
+ * belonging to the same row appear consecutive in the memory and the values of
+ * each batch item are also stored consecutively in memory).
+ *
+ * @note Though the storage layout is the same as the multi-vector object, the
+ * class semantics and the operations it aims to provide are different. Hence it
+ * is recommended to create multi-vector objects if the user means to view the
+ * data as a set of vectors.
+ *
+ * @tparam ValueType  precision of matrix elements
+ *
+ * @ingroup batch_dense
+ * @ingroup mat_formats
+ * @ingroup BatchLinOp
+ */
+template <typename ValueType = default_precision>
+class Dense final : public EnableBatchLinOp<Dense<ValueType>>,
+                    public EnableCreateMethod<Dense<ValueType>>,
+                    public ConvertibleTo<Dense<next_precision<ValueType>>> {
+    friend class EnableCreateMethod<Dense>;
+    friend class EnablePolymorphicObject<Dense, BatchLinOp>;
+    friend class Dense<to_complex<ValueType>>;
+    friend class Dense<next_precision<ValueType>>;
+
+public:
+    using EnableBatchLinOp<Dense>::convert_to;
+    using EnableBatchLinOp<Dense>::move_to;
+
+    using value_type = ValueType;
+    using index_type = int32;
+    using transposed_type = Dense<ValueType>;
+    using unbatch_type = gko::matrix::Dense<ValueType>;
+    using absolute_type = remove_complex<Dense>;
+    using complex_type = to_complex<Dense>;
+
+    /**
+     * Creates a Dense matrix with the configuration of another Dense
+     * matrix.
+     *
+     * @param other  The other matrix whose configuration needs to copied.
+     */
+    static std::unique_ptr<Dense> create_with_config_of(
+        ptr_param<const Dense> other);
+
+    void convert_to(Dense<next_precision<ValueType>>* result) const override;
+
+    void move_to(Dense<next_precision<ValueType>>* result) override;
+
+    /**
+     * Creates a mutable view (of gko::matrix::Dense type) of one item of the
+     * batch::matrix::Dense<value_type> object. Does not perform any deep
+     * copies, but only returns a view of the data.
+     *
+     * @param item_id  The index of the batch item
+     *
+     * @return  a gko::matrix::Dense object with the data from the batch item
+     * at the given index.
+     */
+    std::unique_ptr<unbatch_type> create_view_for_item(size_type item_id);
+
+    /**
+     * @copydoc create_view_for_item(size_type)
+     */
+    std::unique_ptr<const unbatch_type> create_const_view_for_item(
+        size_type item_id) const;
```

Comment:
Somewhat unrelated, but everywhere in Ginkgo, do we actually need to have different names for the const version? Usually the compiler does a good job at selecting the correct version with const overloading.
https://isocpp.org/wiki/faq/const-correctness#const-overloading

The stdlib also usually has the same method for both const & non const access, see e.g.:
https://en.cppreference.com/w/cpp/ranges/view_interface/data
https://en.cppreference.com/w/cpp/container/vector/data

I see that we never mentioned this in either #402 or #1400

Reply:
I think we discussed this sometime back and decided that for uniformity, we have named functions for const qualified related operations. I think this was around the time we were creating `const_view` in `gko::array`